sdks/python/apache_beam/tools/coders_microbenchmark.py

```diff
@@ -16,126 +16,129 @@
+how the coders are used in PCollections: we encode and decode
+a (large) list of (small) collections. A collection can be a list of ints,
+a tuple of windowed values, etc.
+
+In this suite, an element of a benchmark is a collection with several entries.
+
+Run as:
   python -m apache_beam.tools.coders_microbenchmark
+
 """
 
 from __future__ import absolute_import
 from __future__ import print_function
 
-import collections
+import math
 import random
 import string
-import sys
 
 from apache_beam.coders import coders
 from apache_beam.tools import utils
 from apache_beam.transforms import window
 from apache_beam.utils import windowed_value
 
+NUM_ENTRIES_PER_COLLECTION = 50
+
+
+def coder_benchmark_factory(coder, generate_fn):
+  """ Creates a benchmark that encodes and decodes a list of collections.
+
+  Args:
+    coder: coder to use to encode individual collections.
+    generate_fn: a callable that generates a collection of a given size.
+  """
 
-def coder_benchmark_factory(name, coder, generate_fn):
-  def init(self, size):
-    self.coder_ = coder
-    self.data_ = generate_fn(size)
+  class CoderBenchmark(object):
+    def __init__(self, num_collections_per_benchmark):
+      self._coder = coders.IterableCoder(coder)
+      self._list = [generate_fn(NUM_ENTRIES_PER_COLLECTION)
+                    for _ in range(num_collections_per_benchmark)]
 
-  def call(self):
-    _ = self.coder_.decode(self.coder_.encode(self.data_))
+    def __call__(self):
+      _ = self._coder.decode(self._coder.encode(self._list))
 
-  benchmark = type(name, (object,), {})
-  benchmark.__init__ = init
-  benchmark.__call__ = call
-  return benchmark
+  CoderBenchmark.__name__ = "%s(%d entries), %s" % (
+      generate_fn.__name__, NUM_ENTRIES_PER_COLLECTION, str(coder))
 
+  return CoderBenchmark
 
-def generate_list_of_ints(input_size, lower_bound=0, upper_bound=sys.maxsize):
-  values = []
-  for _ in range(input_size):
-    values.append(random.randint(lower_bound, upper_bound))
-  return values
 
+def list_of_ints(size, upper_bound=1000000):
+  """Generates a list of numbers with a distribution skewed to small values."""
+  return [int(math.exp(random.random()**3*math.log(upper_bound)))
+          for _ in range(size)]
 
-def generate_tuple_of_ints(input_size):
-  return tuple(generate_list_of_ints(input_size))
 
+def tuple_of_ints(size):
+  return tuple(list_of_ints(size))
 
-def generate_string(length):
+
+def random_string(length):
   return unicode(''.join(random.choice(
       string.ascii_letters + string.digits) for _ in range(length)))
 
 
-def generate_dict_int_int(input_size):
-  sample_list = generate_list_of_ints(input_size)
-  return {val: val for val in sample_list}
+def dict_int_int(size):
+  return {i: i for i in list_of_ints(size)}
 
 
-def generate_dict_str_int(input_size):
-  sample_list = generate_list_of_ints(input_size)
-  return {generate_string(100): val for val in sample_list}
+def dict_str_int(size):
```

Comment:
It seems for here (and elsewhere) size is always equal to NUM_ENTRIES_PER_COLLECTION. 

It seems important to vary this parameter, e.g. if I were to list the things I wanted to benchmark it would be something like

small ints
large ints
small strings
large strings
small dicts
large dicts
small lists/tuples - 2 tuples are particularly common
large lists/tuples
windowed values with one window
windowed values with more than one window
...

On the other hand, dict_int_int vs dict_str_str probably isn't going to tell us much about the performance of dict.